proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+##### Use a non-escapable index type
+Eventually we want a similar usage pattern for a `MutableSpan` as we are proposing for `Span`. If the index of a `MutableSpan` were to borrow the view, then it becomes impossible to implement a mutating subscript without also requiring an index to be consumed. This seems untenable.
+
+##### Naming
+
+The ideas in this proposal previously used the name `BufferView`. While the use of the word "buffer" would be consistent with the `UnsafeBufferPointer` type, it is nevertheless not a great name, since "buffer" is usually used in reference to transient storage. On the other hand we already have a nomenclature using the term "Storage" in the `withContiguousStorageIfAvailable()` function, and we tried to allude to that in a previous pitch where we called this type `StorageView`. We also considered the name `StorageSpan`, but that did not add much beyond the name `Span`. `Span` clearly identifies itself as a relative of C++'s `std::span`.
+
+##### Adding `load` and `loadUnaligned` to `Span<UInt8>`on `Span<some BitwiseCopyable>` instead of adding `RawSpan`
+
+TKTKTK
+
+##### A more sophisticated approach to indexing
+
+This is discussed more fully in the [indexing appendix](#Indexing) below.
+
+## Future directions
+
+##### Defining `BorrowingIterator` with support in `for` loops
+This proposal does not define an `IteratorProtocol` conformance, since it would need to be borrowed and non-escapable. This is not compatible with `IteratorProtocol`. As such, `Span`  is not directly usable in `for` loops as currently defined. A `BorrowingIterator` protocol for non-escapable and non-copyable containers must be defined, providing a `for` loop syntax where the element is borrowed through each iteration. Ultimately we should arrive at a way to iterate through borrowed elements from a borrowed view:
+
+```swift
+borrowing view: Span<Element> = ...
+for borrowing element in view {
+  doSomething(element)
+}
+```
+
+In the meantime, it is possible to loop through a `Span`'s elements by direct indexing:
+
+```swift
+func doSomething(_ e: borrowing Element) { ... }
+let view: Span<Element> = ...
+// either:
+var i = 0
+while i < view.count {
+  doSomething(view[i])
+  view.index(after: &i)
+}
+
+// ...or:
+for i in 0..<view.indices {
+  doSomething(view[i])
+}
+```
+
+##### Collection-like protocols for non-copyable and non-escapable types
+
+Non-copyable and non-escapable containers would benefit from a `Collection`-like protocol family to represent a set basic, common operations. This may be `Collection` if we find a way to make it work; it may be something else.
```

Comment:
Let's not get people's hopes up. `Collection` is too deeply tied to copyability; we aren't working on its generalization (neither for noncopyable/nonescapable conforming types, nor noncopyable Elements).

Reply:
Removed the second sentence of this paragraph.